### PR TITLE
Wip/version bumps

### PR DESCRIPTION
**Pull request by s-jse**
Version upgrades, general improvements and some nice-to-have features:
- Upgrade `transformers` to version 4.1.1.
- Upgrade `Pytorch` to version 1.7.1. Notably, this version includes this fix https://github.com/pytorch/pytorch/pull/48936 which is important for  mixed precision training.
- (bug fix) `TransformerSeq2Seq` models now use `attention_mask` properly.
- TransformerSeq2Seq models now average loss over actual sequence lengths, **excluding** pad tokens. My experiments show that this way, average log probabilities are better calibrated.
- Support "Diverse Beam Search" (https://arxiv.org/abs/1610.02424) decoding algorithm introduced in the new `transformers` version.
- Add the "Decoupled Weight Decay Regularization" (AdamW) (https://arxiv.org/abs/1711.05101) optimization algorithm.
- Add Loss Truncation (https://arxiv.org/abs/2004.14589)
- Add linear and constant learning rate schedules. Now all schedules and optimizers use `--lr_schedule` and `--lr_multiply`. No more `--transformer_lr` and `--transformer_lr_multiply`.
- Add batch encoding and tokenization using multiprocessing, and if `FastTokenizer`s are used, multithreading. This way, a 1 GB dataset is preprocessed and tokenized in only ~11 minutes on an 8-core CPU, instead of hours.
- Improved batching. Now we make sure batches never exceed the specified batch size.
- Change secondary metric of parsing tasks to F1 score instead of BLEU. BLEU is normally used for natural language tasks, not code generation.
- Improved logging messages.
- Remove some unused code.

### CODE REVIEW

File: genienlp/arguments.py [hunk 1]
```diff
@@ -192,10 +195,8 @@ def post_parse(args):
         for t in args.train_task_names:
             if t not in args.val_task_names:
                 args.val_task_names.append(t)
-    if 'imdb' in args.val_task_names:
-        args.val_task_names.remove('imdb')
 
-    args.timestamp = datetime.datetime.now(tz=datetime.timezone.utc).isoformat()
+    args.timestamp = datetime.datetime.now(tz=datetime.timezone.utc).strftime('%D-%H:%M:%S %Z')
```
Comment (gcampax) [inline genienlp/arguments.py:201]: Why these two changes?
(imdb does not have validation data so the previous change was correct)

File: genienlp/tasks/almond/__init__.py [hunk 2]
```diff
@@ -288,9 +292,27 @@ def _make_example(self, parts, dir_name=None, **kwargs):
...
+    def preprocess_field(self, sentence, field_name=None):
+        # remove the $ at the start of dialogue
+        # this is safe because we know we're processing dialogues, so the answer
+        # always starts with $dialogue and the context is either `null` or also
+        # starts with $dialogue
+        if field_name == 'context' and sentence.startswith('$dialogue @org.thingpedia.dialogue.transaction . '):
+            sentence = sentence[len('$dialogue @org.thingpedia.dialogue.transaction . '):]
+        if field_name == 'answer':
+            assert(sentence.startswith('$dialogue @org.thingpedia.dialogue.transaction . '))
+            sentence = sentence[len('$dialogue @org.thingpedia.dialogue.transaction . '):]
```
Comment (gcampax) [inline genienlp/tasks/almond/__init__.py]: Removing `$dialogue` is safe, removing `@org.thingpedia.dialogue.transaction` is not safe, because there might other dialogue policies in the future. Also, if you remove `$dialogue` entirely, you should update the comment.

File: genienlp/arguments.py (hunk 1 above)
Comment (s-jse) [reply genienlp/arguments.py:201]: I added the imdb back.
The `strftime` formatting reduces the width of log messages a bit.

File: genienlp/tasks/almond/__init__.py (hunk 2 above)
Comment (s-jse) [reply genienlp/tasks/almond/__init__.py]: OK, I changed it to `$dialogue`.

**Comment by s-jse** [comment]
Can we merge this now?

**Comment by gcampax** [comment]
Yeah, let's merge this.
